nucleus/iou_match.py

```diff
@@ -0,0 +1,75 @@
+from dataclasses import dataclass
+from enum import Enum
+
+from .constants import (
+    DATASET_ITEM_ID_KEY,
+    GROUND_TRUTH_ANNOTATION_ID_KEY,
+    GROUND_TRUTH_ANNOTATION_LABEL_KEY,
+    IOU_KEY,
+    MODEL_PREDICTION_ID_KEY,
+    MODEL_PREDICTION_LABEL_KEY,
+    MODEL_RUN_ID_KEY,
+)
+
+
+class ConfusionCategory(Enum):
+    TRUE_POSITIVE = "true_positive"
+    TRUE_NEGATIVE = "true_negative"
+    FALSE_POSITIVE = "false_positive"
+    FALSE_NEGATIVE = "false_negative"
+
+
+def infer_confusion_category(
+    ground_truth_annotation_label: str,
+    model_prediction_label: str,
+):
+    confusion_category = ConfusionCategory.FALSE_NEGATIVE
+
+    if (
+        model_prediction_label is None
+        and ground_truth_annotation_label is None
+    ):
+        confusion_category = ConfusionCategory.TRUE_NEGATIVE
+    elif model_prediction_label == ground_truth_annotation_label:
+        confusion_category = ConfusionCategory.TRUE_POSITIVE
+    elif model_prediction_label is not None:
+        confusion_category = ConfusionCategory.FALSE_POSITIVE
+
+    return confusion_category
+
+
+# TODO: Is there a better name for this class?
+@dataclass
+class IOUMatch:
+    model_run_id: str
+    model_prediction_id: str  # field is nullable
+    ground_truth_annotation_id: str  # field is nullable
+    iou: int
+    dataset_item_id: str
+    confusion_category: ConfusionCategory
+    model_prediction_label: str  # field is nullable
+    ground_truth_annotation_label: str  # field is nullable
+
+    @classmethod
+    def from_json(cls, payload: dict):
+        model_prediction_label = payload.get(MODEL_PREDICTION_LABEL_KEY, None)
+        ground_truth_annotation_label = payload.get(
+            GROUND_TRUTH_ANNOTATION_LABEL_KEY, None
+        )
+        confusion_category = infer_confusion_category(
+            ground_truth_annotation_label=ground_truth_annotation_label,
+            model_prediction_label=model_prediction_label,
+        )
```

Comment:
Maybe this was a bit of a confusing ask: we should propogate the true_positive to the SDK but then use that to map the confusion type. 🙂